src/client/gui/lib/vm_table/bulk_actions.dart

```diff
@@ -9,6 +9,126 @@ import '../providers.dart';
+  const ZonesDropdown({super.key});
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    // Pre-fetch zones data
+    final zonesAsync = ref.watch(zonesProvider);
+    final vmInfos = ref.watch(vmInfosProvider);
+    return PopupMenuButton(
+      tooltip: 'Change zone availability',
+      position: PopupMenuPosition.under,
+      child: OutlinedButton(
+        onPressed: null,
+        child: const Text('Zones', style: TextStyle(color: Colors.black)),
+      ),
+      itemBuilder: (context) => zonesAsync.when(
+          loading: () => [
+                const PopupMenuItem(
+                  enabled: false,
+                  child: SizedBox(
+                    width: 20,
+                    height: 20,
+                    child: CircularProgressIndicator(),
+                  ),
+                ),
+              ],
+          error: (_, __) => [
+                const PopupMenuItem(
+                  enabled: false,
+                  child: Text('Error loading zones'),
+                ),
+              ],
+          data: (zones) => [
+                const PopupMenuItem(
+                  enabled: false,
+                  child: Padding(
+                    padding: EdgeInsets.symmetric(horizontal: 4, vertical: 8),
+                    child: Text(
+                      'Availability zones',
+                      style: TextStyle(
+                        fontSize: 24,
+                        fontWeight: FontWeight.w600,
+                        color: Colors.black,
+                      ),
+                    ),
+                  ),
+                ),
+                ...zones.map((zone) {
+                  // Count running instances in this zone
+                  final runningCount = vmInfos
+                      .where((vm) =>
+                          vm.zone?.name == zone.name &&
+                          vm.instanceStatus.status == Status.RUNNING)
+                      .length;
+
+                  return PopupMenuItem(
+                    enabled: false,
+                    padding: EdgeInsets.zero,
+                    child: Container(
+                      decoration: BoxDecoration(
+                        border: Border.all(color: Colors.grey.withOpacity(0.3)),
+                        borderRadius: BorderRadius.circular(4),
+                      ),
+                      margin: const EdgeInsets.symmetric(
+                          horizontal: 8, vertical: 4),
+                      padding: const EdgeInsets.all(8),
+                      child: _ZoneToggleRow(
+                        zone.name,
+                        zone.available,
+                        runningCount,
+                      ),
+                    ),
+                  );
+                }).toList(),
+              ]),
+    );
+  }
+}
+
+class _ZoneToggleRow extends ConsumerWidget {
+  final String zoneName;
+  final bool available;
+  final int instanceCount;
```

Comment:
I was wrong about my earlier advice of passing those in the constructor. You were right to watch providers here. Since those values are passed in, they won't update when the provider changes, since they are in a popup. So you need to go back to watching the providers directly in this widget, and have this widget receive just the zone name in the constructor. You'll need to add some code like this
```dart
final available = ref.watch(zonesProvider.select((zones) {
  return zones.firstWhere((z) => z.name == zoneName).available;
}));
final runningInstances = ref.watch(vmInfosProvider.select((infos) {
  return infos
      .where((vm) => vm.zone.name == zoneName)
      .where((vm) => vm.instanceStatus.status == Status.RUNNING)
      .length;
}));
```